### DISCUSSION TITLE
Finding URL of request

### DISCUSSION THREAD

**Discussion by docktermj** [ogen-go/ogen#931] [Q&A]
### What version of ogen are you using?

```bash
$ go list -m github.com/ogen-go/ogen

github.com/ogen-go/ogen v0.68.3
```

### Can this issue be reproduced with the latest version?

Yes

### What did you do?

Given the code generated by ogen:

https://github.com/docktermj/go-rest-api-client/blob/ef03165351138463685dc49c57c59670c30b8347/senzingrestapi/oas_unimplemented_gen.go#L331-L333

How do I find the URL of the request?   Specifically the hostname of the requester.

I know it's buried in the context `withValue()`

```
field name: children
value: map[context.Background.WithValue(type *http.contextKey, val <not Stringer>).WithValue(type *http.contextKey, val 127.0.0.1:8261).WithCancel.WithCancel:{}]
```

but I can't find the keys so that I can do 

```go
myUrl, _ := ctx.Value(???).(*.url.URL)
```

Is that how I should be finding the URL?   If so, what is the value of "???" I should use?

**Comment by ernado** [comment]
You can use [nethttp middlewares](https://ogen.dev/docs/concepts/static_router#using-nethttp-middlewares) and put URL (or anything from `*http.Request`) to context.

**Comment by docktermj** [comment]
I'm sorry, I'm not understanding.   I'm not looking to put something into the context.   I'm seeking to find something in the context that was put into the context by ogen.

For instance, in code that implements https://github.com/docktermj/go-rest-api-client/blob/ef03165351138463685dc49c57c59670c30b8347/senzingrestapi/oas_unimplemented_gen.go#L331-L333, would look something like:

```go
func (c MyStruct) Heartbeat(ctx context.Context) (r *SzBaseResponse, _ error) {
        requestUrl := ctx.Value(????).(*url.URL)
        :
}
```

What do I replace "????" to get the `*url.URL` value that ogen put into the context?

**Reply by ghost** [reply]
ogen does not pass any values into context, at least directly (OpenTelemetry tracer possibly could). 

Values you see are probably passed by `net/http`: [`http.LocalAddrContextKey`](https://pkg.go.dev/net/http#LocalAddrContextKey) and [`http.ServerContextKey`](https://pkg.go.dev/net/http#ServerContextKey)

As @ernado pointed out, you could write a middleware and pass any values to request's context.

For example:
```go
type myContextKey struct{}

func WithRequestURL(ctx context.Context, u *url.URL) context.Context {
       return context.WithValue(ctx, myContextKey{}, u)
}

func GetRequestURL(ctx context.Context) (*url.URL, bool) {
      u, ok := ctx.Value(myContextKey{})
      return u, ok
}

func Middleware(h http.Handler) http.Handler {
    return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
        ctx := WithRequestURL(r.Context(), r.URL)
        h.ServeHTTP(w, r.WithContext(ctx))
    })
}

...

func (c MyStruct) Heartbeat(ctx context.Context) (r *SzBaseResponse, _ error) {
        requestURL, ok := GetRequestURL(ctx)
        ...
}
```
